File: Backend/controllers/predictionController.js
const { execFile } = require('child_process');
const fs = require('fs');
const path = require('path');
const Prediction = require('../models/predictionModel');
const Data = require('../models/dataModel');
const PredictionInput = require('../models/predictionInputModel');
const { calculateRiskScore, getMitigationRecommendation } = require('../utils/helpers');

const PYTHON = process.env.PYTHON_PATH || path.join(__dirname, '../../Training/.venv310/Scripts/python.exe');
const PREDICT_SCRIPT = path.join(__dirname, '../../Training/predict.py');
const MODEL_FILE = path.join(__dirname, '../../Training/saved_models/best_gru_model.h5');

const runPythonPrediction = (history, steps) =>
  new Promise((resolve, reject) => {
    execFile(PYTHON, [PREDICT_SCRIPT, JSON.stringify({ history, steps })], { timeout: 120000 }, (err, stdout, stderr) => {
      if (err) return reject(new Error(stderr || err.message));
      try {
        const parsed = JSON.parse(stdout);
        if (!parsed.success) return reject(new Error(parsed.error || 'Prediksi gagal'));
        resolve(parsed.predictions);
      } catch (e) {
        reject(e);
      }
    });
  });

const PredictionController = {
  getAll: async (req, res) => {
    try {
      const data = await Prediction.findAll(req.user.id);
      res.json({ success: true, data });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  getLatest: async (req, res) => {
    try {
      const lokasi = req.query.lokasi || null;
      const data = await Prediction.findLatest(req.user.id, lokasi);
      res.json({ success: true, data });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  getDashboard: async (req, res) => {
    try {
      const predictions = await Prediction.findAll(req.user.id);
      const rank = { Normal: 0, Waspada: 1, Bahaya: 2 };

      const byLokasi = {};
      predictions.forEach((p) => {
        const key = p.lokasi || 'Tanpa Lokasi';
        if (!byLokasi[key]) byLokasi[key] = [];
        byLokasi[key].push(p);
      });

      const perLokasi = Object.keys(byLokasi).map((lokasi) => {
        const rows = byLokasi[lokasi];
        let worst = rows[0];
        rows.forEach((r) => {
          if (
            rank[r.status] > rank[worst.status] ||
            (rank[r.status] === rank[worst.status] && Number(r.skor_risiko) < Number(worst.skor_risiko))
          ) {
            worst = r;
          }
        });
        return {
          lokasi,
          suhu: worst.suhu,
          pH: worst.pH,
          salinitas: worst.salinitas,
          kekeruhan: worst.kekeruhan,
          skor_risiko: worst.skor_risiko,
          status: worst.status,
          rekomendasi: worst.rekomendasi
        };
      });

      const n = predictions.length;
      const avg = (key) => (n ? predictions.reduce((s, p) => s + parseFloat(p[key] || 0), 0) / n : 0);
      const stats = {
        total_data: n,
        avg_suhu: avg('suhu'),
        avg_pH: avg('pH'),
        avg_salinitas: avg('salinitas'),
        avg_kekeruhan: avg('kekeruhan'),
        normal_count: predictions.filter((p) => p.status === 'Normal').length,
        waspada_count: predictions.filter((p) => p.status === 'Waspada').length,
        bahaya_count: predictions.filter((p) => p.status === 'Bahaya').length
      };

      const maxSteps = Math.max(0, ...Object.values(byLokasi).map((a) => a.length));
      const series = [];
      for (let i = 1; i <= maxSteps; i++) {
        const steps = predictions.filter((p) => Number(p.step_ke) === i);
        if (!steps.length) continue;
        const avg2 = (key) => steps.reduce((s, p) => s + parseFloat(p[key] || 0), 0) / steps.length;
        series.push({
          step_ke: i,
          suhu: +avg2('suhu').toFixed(2),
          pH: +avg2('pH').toFixed(2),
          salinitas: +avg2('salinitas').toFixed(2),
          kekeruhan: +avg2('kekeruhan').toFixed(2)
        });
      }

      res.json({ success: true, data: { stats, perLokasi, series } });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  runPrediction: async (req, res) => {
    try {
      let { suhu, pH, salinitas, kekeruhan, waktu, lokasi } = req.body;
      lokasi = lokasi || null;

      if (suhu === undefined || pH === undefined || salinitas === undefined || kekeruhan === undefined) {
        const latest = await Data.findLatest(req.user.id, lokasi);
        if (!latest) return res.status(400).json({ success: false, message: 'Tidak ada data terbaru. Isi nilai parameter manual.' });
        suhu = latest.suhu;
        pH = latest.pH;
        salinitas = latest.salinitas;
        kekeruhan = latest.kekeruhan;
      }

      suhu = parseFloat(suhu);
      pH = parseFloat(pH);
      salinitas = parseFloat(salinitas);
      kekeruhan = parseFloat(kekeruhan);

      if ([suhu, pH, salinitas, kekeruhan].some((v) => !Number.isFinite(v))) {
        return res.status(400).json({ success: false, message: 'Semua parameter (suhu, pH, salinitas, kekeruhan) harus diisi dengan angka yang valid' });
      }

      const inputParams = { suhu, pH, salinitas, kekeruhan };
      const startDate = waktu ? new Date(waktu) : new Date();

      await Prediction.deleteAll(req.user.id, lokasi);

      // Bangun window 24 timestep
      let window;
      const inputProvided = req.body.suhu !== undefined && req.body.pH !== undefined && req.body.salinitas !== undefined && req.body.kekeruhan !== undefined;

      if (inputProvided) {
        // User mengisi nilai manual: window mengikuti nilai input, bukan riwayat DB
        window = Array.from({ length: 24 }, () => [suhu, pH, salinitas, kekeruhan]);
      } else {
        // Tidak mengisi: gunakan riwayat terbaru DB
        const historyRows = await Data.findAll(23, 0, req.user.id, lokasi);
        historyRows.reverse();
        window = historyRows.map((r) => [r.suhu, r.pH, r.salinitas, r.kekeruhan]);
        while (window.length < 23) window.unshift([suhu, pH, salinitas, kekeruhan]);
        window.push([suhu, pH, salinitas, kekeruhan]);
      }

      let rawPredictions = null;
      if (fs.existsSync(PYTHON) && fs.existsSync(PREDICT_SCRIPT) && fs.existsSync(MODEL_FILE)) {
        try {
          rawPredictions = await runPythonPrediction(window, 96);
          console.log('Prediksi menggunakan model GRU (PSO)');
        } catch (error) {
          console.error('Python prediction error, fallback ke simulasi:', error.message);
        }
      } else {
        console.log('Model GRU tidak ditemukan, fallback ke simulasi');
      }

      // Simulasi prediksi 96 langkah (15 menit, 1 hari ke depan)
      const predictions = [];
      for (let i = 1; i <= 96; i++) {
        if (rawPredictions && rawPredictions[i - 1]) {
          suhu = +rawPredictions[i - 1][0].toFixed(2);
          pH = +rawPredictions[i - 1][1].toFixed(2);
          salinitas = +rawPredictions[i - 1][2].toFixed(2);
          kekeruhan = +rawPredictions[i - 1][3].toFixed(2);
        } else {
          suhu = +(suhu + (Math.random() - 0.5) * 0.2).toFixed(2);
          pH = +(pH + (Math.random() - 0.5) * 0.03).toFixed(2);
          salinitas = +(salinitas + (Math.random() - 0.5) * 0.15).toFixed(2);
          kekeruhan = +(kekeruhan + (Math.random() - 0.5) * 0.15).toFixed(2);
          suhu = Math.min(32, Math.max(26, suhu));
          pH = Math.min(8.5, Math.max(7.0, pH));
          salinitas = Math.min(35, Math.max(10, salinitas));
          kekeruhan = Math.min(5, Math.max(0, kekeruhan));
        }

        const risk = calculateRiskScore(suhu, pH, salinitas, kekeruhan);
        const rekomendasi = getMitigationRecommendation(risk.status);

        const date = new Date(startDate.getTime() + (i - 1) * 15 * 60 * 1000);
        const pad2 = (n) => String(n).padStart(2, '0');
        const localDate =
          date.getFullYear() + '-' +
          pad2(date.getMonth() + 1) + '-' +
          pad2(date.getDate()) + ' ' +
          pad2(date.getHours()) + ':' +
          pad2(date.getMinutes()) + ':' +
          pad2(date.getSeconds());
        predictions.push({
          user_id: req.user.id,
          lokasi,
          tanggal_prediksi: localDate,
          step_ke: i,
          suhu,
          pH,
          salinitas,
          kekeruhan,
          skor_risiko: risk.skor,
          status: risk.status,
          rekomendasi,
          model_log_id: null
        });
      }

      await Prediction.createBatch(predictions);
      await PredictionInput.create({
        user_id: req.user.id,
        lokasi,
        tanggal_prediksi: startDate,
        nilai_parameter: inputParams
      });

      const saved = await Prediction.findLatest(req.user.id, lokasi);
      res.json({ success: true, data: saved });
    } catch (error) {
      console.error('Prediction error:', error.message);
      res.status(500).json({ success: false, message: 'Server error: ' + error.message });
    }
  }
};

module.exports = PredictionController;

File: backend_python/controllers/prediction_controller.py
import json
import os
import random
import subprocess
from datetime import datetime, timedelta

from flask import g, request

from models.data_model import Data
from models.prediction_input_model import PredictionInput
from models.prediction_model import Prediction
from utils.helpers import calculate_risk_score, fail, finite, get_mitigation_recommendation, ok

PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
PYTHON = os.getenv("PYTHON_PATH") or os.path.join(PROJECT_ROOT, "Training", ".venv310", "Scripts", "python.exe")
PREDICT_SCRIPT = os.path.join(PROJECT_ROOT, "Training", "predict.py")
MODEL_FILE = os.path.join(PROJECT_ROOT, "Training", "saved_models", "best_gru_model.h5")

RANK = {"Normal": 0, "Waspada": 1, "Bahaya": 2}


def _run_python_prediction(window, steps):
    proc = subprocess.run(
        [PYTHON, PREDICT_SCRIPT, json.dumps({"history": window, "steps": steps})],
        timeout=120,
        capture_output=True,
        text=True,
    )
    if proc.returncode != 0:
        raise RuntimeError(proc.stderr or "Prediksi gagal")
    parsed = json.loads(proc.stdout)
    if not parsed.get("success"):
        raise RuntimeError(parsed.get("error") or "Prediksi gagal")
    return parsed["predictions"]


def _fmt_local(dt):
    return dt.strftime("%Y-%m-%d %H:%M:%S")


class PredictionController:
    @staticmethod
    def get_all():
        return ok({"success": True, "data": Prediction.find_all(g.user["id"])})

    @staticmethod
    def get_latest():
        lokasi = request.args.get("lokasi") or None
        return ok({"success": True, "data": Prediction.find_latest(g.user["id"], lokasi)})

    @staticmethod
    def get_dashboard():
        predictions = Prediction.find_all(g.user["id"])

        by_lokasi = {}
        for p in predictions:
            key = p.get("lokasi") or "Tanpa Lokasi"
            by_lokasi.setdefault(key, []).append(p)

        per_lokasi = []
        for lokasi, rows in by_lokasi.items():
            worst = rows[0]
            for r in rows[1:]:
                if RANK.get(r["status"], 3) > RANK.get(worst["status"], 3) or (
                    RANK.get(r["status"], 3) == RANK.get(worst["status"], 3)
                    and float(r["skor_risiko"]) < float(worst["skor_risiko"])
                ):
                    worst = r
            per_lokasi.append(
                {
                    "lokasi": lokasi,
                    "suhu": worst["suhu"],
                    "pH": worst["pH"],
                    "salinitas": worst["salinitas"],
                    "kekeruhan": worst["kekeruhan"],
                    "skor_risiko": worst["skor_risiko"],
                    "status": worst["status"],
                    "rekomendasi": worst["rekomendasi"],
                }
            )

        n = len(predictions)

        def avg(key, items=None):
            items = items if items is not None else predictions
            return sum(float(p[key] or 0) for p in items) / n if n else 0

        stats = {
            "total_data": n,
            "avg_suhu": avg("suhu"),
            "avg_pH": avg("pH"),
            "avg_salinitas": avg("salinitas"),
            "avg_kekeruhan": avg("kekeruhan"),
            "normal_count": sum(1 for p in predictions if p["status"] == "Normal"),
            "waspada_count": sum(1 for p in predictions if p["status"] == "Waspada"),
            "bahaya_count": sum(1 for p in predictions if p["status"] == "Bahaya"),
        }

        max_steps = max((len(v) for v in by_lokasi.values()), default=0)
        series = []
        for i in range(1, max_steps + 1):
            steps = [p for p in predictions if int(p["step_ke"]) == i]
            if not steps:
                continue

            def avg2(key):
                return sum(float(p[key] or 0) for p in steps) / len(steps)

            series.append(
                {
                    "step_ke": i,
                    "suhu": round(avg2("suhu"), 2),
                    "pH": round(avg2("pH"), 2),
                    "salinitas": round(avg2("salinitas"), 2),
                    "kekeruhan": round(avg2("kekeruhan"), 2),
                }
            )

        return ok({"success": True, "data": {"stats": stats, "perLokasi": per_lokasi, "series": series}})

    @staticmethod
    def run_prediction():
        body = request.get_json(silent=True) or {}
        suhu, ph = body.get("suhu"), body.get("pH")
        salinitas, kekeruhan = body.get("salinitas"), body.get("kekeruhan")
        waktu, lokasi = body.get("waktu"), body.get("lokasi") or None

        if suhu is None or ph is None or salinitas is None or kekeruhan is None:
            latest = Data.find_latest(g.user["id"], lokasi)
            if not latest:
                return fail("Tidak ada data terbaru. Isi nilai parameter manual.", 400)
            suhu, ph = latest["suhu"], latest["pH"]
            salinitas, kekeruhan = latest["salinitas"], latest["kekeruhan"]

        try:
            suhu, ph = float(suhu), float(ph)
            salinitas, kekeruhan = float(salinitas), float(kekeruhan)
        except (TypeError, ValueError):
            return fail("Semua parameter (suhu, pH, salinitas, kekeruhan) harus diisi dengan angka yang valid", 400)
        if not finite(suhu, ph, salinitas, kekeruhan):
            return fail("Semua parameter (suhu, pH, salinitas, kekeruhan) harus diisi dengan angka yang valid", 400)

        input_params = {"suhu": suhu, "pH": ph, "salinitas": salinitas, "kekeruhan": kekeruhan}
        try:
            start_date = datetime.fromisoformat(str(waktu)) if waktu else datetime.now()
        except ValueError:
            start_date = datetime.now()

        Prediction.delete_all(g.user["id"], lokasi)

        input_provided = all(k in body and body[k] is not None for k in ("suhu", "pH", "salinitas", "kekeruhan"))
        if input_provided:
            window = [[suhu, ph, salinitas, kekeruhan]] * 24
        else:
            history_rows = Data.find_all(23, 0, g.user["id"], lokasi)
            history_rows.reverse()
            window = [[r["suhu"], r["pH"], r["salinitas"], r["kekeruhan"]] for r in history_rows]
            while len(window) < 23:
                window.insert(0, [suhu, ph, salinitas, kekeruhan])
            window.append([suhu, ph, salinitas, kekeruhan])

        raw_predictions = None
        if os.path.exists(PYTHON) and os.path.exists(PREDICT_SCRIPT) and os.path.exists(MODEL_FILE):
            try:
                raw_predictions = _run_python_prediction(window, 96)
                print("Prediksi menggunakan model GRU (PSO)")
            except Exception as e:
                print(f"Python prediction error, fallback ke simulasi: {e}")
        else:
            print("Model GRU tidak ditemukan, fallback ke simulasi")

        predictions = []
        cur_suhu, cur_ph = suhu, ph
        cur_sal, cur_turb = salinitas, kekeruhan
        for i in range(1, 97):
            if raw_predictions and raw_predictions[i - 1]:
                cur_suhu = round(float(raw_predictions[i - 1][0]), 2)
                cur_ph = round(float(raw_predictions[i - 1][1]), 2)
                cur_sal = round(float(raw_predictions[i - 1][2]), 2)
                cur_turb = round(float(raw_predictions[i - 1][3]), 2)
            else:
                cur_suhu = round(cur_suhu + (random.random() - 0.5) * 0.2, 2)
                cur_ph = round(cur_ph + (random.random() - 0.5) * 0.03, 2)
                cur_sal = round(cur_sal + (random.random() - 0.5) * 0.15, 2)
                cur_turb = round(cur_turb + (random.random() - 0.5) * 0.15, 2)
                cur_suhu = min(32, max(26, cur_suhu))
                cur_ph = min(8.5, max(7.0, cur_ph))
                cur_sal = min(35, max(10, cur_sal))
                cur_turb = min(5, max(0, cur_turb))

            risk = calculate_risk_score(cur_suhu, cur_ph, cur_sal, cur_turb)
            rekomendasi = get_mitigation_recommendation(risk["status"])
            date_i = start_date + timedelta(minutes=15 * (i - 1))
            predictions.append(
                {
                    "user_id": g.user["id"],
                    "lokasi": lokasi,
                    "tanggal_prediksi": _fmt_local(date_i),
                    "step_ke": i,
                    "suhu": cur_suhu,
                    "pH": cur_ph,
                    "salinitas": cur_sal,
                    "kekeruhan": cur_turb,
                    "skor_risiko": risk["skor"],
                    "status": risk["status"],
                    "rekomendasi": rekomendasi,
                    "model_log_id": None,
                }
            )

        Prediction.create_batch(predictions)
        PredictionInput.create(
            {
                "user_id": g.user["id"],
                "lokasi": lokasi,
                "tanggal_prediksi": start_date,
                "nilai_parameter": input_params,
            }
        )
        saved = Prediction.find_latest(g.user["id"], lokasi)
        return ok({"success": True, "data": saved})
